### ISSUE TITLE
Hydration implementation causes reconciliation to replace entire DOM sub-tree

### ISSUE THREAD

**Issue by anishathalye** [hashicorp/next-mdx-remote#118]
`hydrate()` behaves poorly with respect to React DOM reconciliation. I'm not providing a minimal demo here, because this happens with any use of next-mdx-remote in the intended way, e.g. see [mdnext-blog](https://github.com/domitriusclark/mdnext-blog) as a small example. If a more self-contained demo would be helpful, please let me know.

Currently, `renderToString()` produces an object with a `compiledSource` (that can be `eval`ed to give a React component hierarchy) and a `renderedOutput` that is plain HTML. The [current implementation](https://github.com/hashicorp/next-mdx-remote/blob/7eaa346348c097d6750685a656bd25440a7e05f4/hydrate.js) of `hydrate()` behaves as follows: on the server, it returns a div with `dangerouslySetInnerHTML` set to the `renderedOutput`, whereas on the client it also does the same thing, but asynchronously evals `compiledSource` to produce a component hierarchy (with some extra details, e.g. wrapping nodes with `MDXContent`, adding a `MDXProvider` at the top of the hierarchy, ...) and replaces the div when it's ready.

The way this ends up working out, the browser loads the initial HTML, then it loads JS, which produces a React element with the dangerouslySetInnerHTML, and then soon after, that element replaces its contents with the real component hierarchy. Even though that component hierarchy renders to the same HTML that's the contents of the div with dangerouslySetInnerHTML, [React's reconciliation](https://reactjs.org/docs/reconciliation.html) ends up removing and replacing the div, because the elements have different types / the component hierarchies are entirely different (single div vs the "real" hierarchy). You can verify this by using the browser to select the element containing the markdown and setting "Break on subtree modification.

This results in bad performance and bad UX. It's not too bad for a simple document just containing e.g. text, but it results in visible flickers when there is fancier stuff going on, e.g. when the document contains a video.

It also results in missing warnings where you might otherwise get them, e.g. when the server's HTML doesn't match the client's rendering. For example, using the following component somewhere else in the Next.js document produces this warning, but using this in the mdx document rendered with next-mdx-remote does not:

```javascript
const Sneaky = () => (
  <p>{typeof window === 'undefined' ? 'server' : 'client'}</p>
)
```

I don't have a deep understanding of the next-mdx-remote implementation, but why can't `hydrate()` simply synchronously render the document, _both on the client and on the server_? Next.js is doing its own rendering to a string / hydration anyways, and this seems to interop nicer with that, e.g. using the above `Sneaky` component in mdx does produce a warning.

I'm currently using the following replacement for next-mdx-remote's hydrate:

```javascript
const hydrate = ({ compiledSource, scope }, { components, provider }) => {
  const fullScope = Object.assign({ mdx: MDX.mdx }, components, scope)
  const keys = Object.keys(fullScope)
  const values = Object.values(fullScope)
  const hydrateFn = Reflect.construct(
    Function,
    ['React']
      .concat(keys)
      .concat(compiledSource + '\nreturn React.createElement(MDXContent, {});')
  )
  const hydrated = hydrateFn.apply(hydrateFn, [React].concat(values))
  const wrappedWithMdxProvider = React.createElement(
    MDX.MDXProvider,
    { components },
    hydrated
  )
  if (provider) {
    return React.createElement(provider.component, provider.props || {}, result)
  }
  return wrappedWithMdxProvider
}
```

It's basically the existing implementation, but synchronously rendering to a React component hierarchy on both the server and the client.

Is this behavior (of React reconciliation replacing the entire DOM subtree corresponding to the markdown output) considered a bug? And if so, is the snippet above the right fix? If that's the case, I'd be happy to clean it up / rewrite it in the more compatible dialect of JS currently used in the existing code, optimize `renderToString` if applicable (producing `renderedOutput` is now redundant, if nothing else depends on it), and send a PR.

**Comment by BRKalow** [comment]
Thanks for the report! I think there is certainly opportunity to cleanup the implementation here, but the way it is functioning now is deliberate. The approach we are using to asynchronously hydrate on the client is done to effectively optimize the initial hydration of the full page, see this note in the hydrate source:

```ts
  // if we're on the client side, we hydrate the mdx content inside
  // requestIdleCallback, since we can be fairly confident that
  // markdown - embedded components are not a high priority to get
  // to interactive compared to...anything else on the page.
```

This approach is similar to the approach sometimes referenced as lazy hydration, and uses the workaround [described here](https://github.com/facebook/react/issues/10923#issuecomment-338715787). Also see the [`react-lazy-hydration`](https://www.npmjs.com/package/react-lazy-hydration) package for a more generic implementation. Hopefully this approach will become redundant when concurrent mode lands in React, but for now it can have measurable positive impact on the performance of the initial page load.

I think one thing we can do is not ship the rendered markup to the client as data, like you said, and instead rely on React's behavior of not touching subtrees wrapped in `dangerouslySetInnerHTML` during hydration. I can definitely see how this could be disruptive for something like a video though, where the element gets removed and re-inserted shortly after the initial page load. Perhaps it's something we can support via an option, or an alternative hydration API 💭... async or sync hydration. That might require further discussion though.

Thanks for opening this, I think we have some opportunity for improvement here!

**Comment by anishathalye** [comment]
That makes sense. In my case, I didn't mind the hydration being lazy per se (I wouldn't mind it being synchronous either), I just didn't want React reconciliation to destroy and re-create that DOM subtree. I think just making hydration synchronous on the client wouldn't fix the problem, if the server-side code stays the way it is (checking if it's running server-side and returning the div with `dangerouslySetInnerHTML`), because then React would (now, synchronously, on the client) replace the DOM subtree. The server-side would also need to produce the component hierarchy including the `MDXProvider`s etc.

I'm not experienced with React, so I don't know what is the right/cleanest way of handling this.

Are there any workarounds for making this behave nicer in situations where there's e.g. a video tag? Searching the issues, I was somewhat surprised that others haven't run into this problem before.

If React had a way to have more fine-grained control over reconciliation for a subtree (e.g. a flag for "expensive mode" reconciliation that does the full O(n^3) minimal tree transformation computation, or a "I am changing the component but I promise the HTML output is the same" flag), that could solve the problem, because the final HTML output corresponding to the `div` with `dangerouslySetInnerHTML` should match the HTML output for the hydrated version.

Would it make sense for next-mdx-remote's `hydrate()`'s options to take a `eager:` or `synchronous:` flag that makes the code render the actual component hierarchy synchronously on both the client and server side (i.e. the behavior of my code snippet above)?

**Comment by peduarte** [comment]
Nice! Thanks for putting this together @anishathalye. I was about to raise an issue related to this too. 

I've been using `next-mdx-remote` on a few projects and have come across an unwelcoming side-effect caused by `questIdleCallback`.

It's quite common that MDX-authored files might have `title`, `published` fields in the frontmatter. And these are typically pulled out and rendered directly on the layout component. Followed by the MDX content. When this is the case, you'll seen an interval between the rendering of the frontmatter data and the mdx content, because the content is in `requestIdleCallback`. 

I've noticed this happening on a few websites using `next-mdx-remote`.

| Radix        | Lee Robinson's Portfolio           | 
| ------------- |-------------|
| <video src="https://user-images.githubusercontent.com/372831/115574438-e76cff00-a2c1-11eb-8aee-3b708215b47c.mp4     "/> | <video src="https://user-images.githubusercontent.com/372831/115574636-1aaf8e00-a2c2-11eb-95a4-a8dc810e318e.mp4"/> |

| Supabase        | Ferrans Portfolio           | 
| ------------- |-------------|
| <video src="https://user-images.githubusercontent.com/372831/115574858-4b8fc300-a2c2-11eb-8a7e-43b9f10634c9.mp4"/>| <video src="https://user-images.githubusercontent.com/372831/115575350-c2c55700-a2c2-11eb-9183-a230d00d29ef.mp4"/> |


We have moved all content into the MDX body. It required us refactoring 90+ mdx files, but that solved the rendering jump. Worth it for the improved user experience.

But we still noticed some rendering flashed sometimes. This mainly happened when using CSS-in-JS, where the delayed rendering was causing a delay of the styles being hydrated too. This happens when some markdown files contain styles that didn't get SSR'ed - therefore needed to be injected client side. A common, and probably unavoidable consequence of CSS critical path rendering.

To fix this, we've ended overriding `window.requestIdleCallback` completely:
```jsx
window.requestIdleCallback = function(cb) {
  return cb()
}
```

Not ideal, but once again, worth it for the improved user experience.

| With `requestIdleCallback`        | Without `requestIdleCallback`           | 
| ------------- |-------------|
| <video src="https://user-images.githubusercontent.com/372831/115576341-b2fa4280-a2c3-11eb-98f5-1b47e5a8229a.mp4"/>| <video src="https://user-images.githubusercontent.com/372831/115576376-ba215080-a2c3-11eb-96ea-da2bb08613f3.mp4"/> |

Hopefully, this has been useful, to better understand how others may consume this library. 

---

I also agree passing an option to disable the async rendering would be awesome! Personally, I'd prefer to make it synchronous by default, with an option to make it async.

Based on this comment:
```
  // if we're on the client side, we hydrate the mdx content inside
  // requestIdleCallback, since we can be fairly confident that
  // markdown - embedded components are not a high priority to get
  // to interactive compared to...anything else on the page.
```

Personally, I'd argue that for a lot of websites using MDX, the MDX content *is the product* and therefore, it is a high priority.

Thanks!

**Comment by BRKalow** [comment]
This has been super helpful, thanks for taking the time both of you!

If y'all are interested, we've got a new major version in the works which I think should solve this problem, fix most CSS-in-JS SSR issues, as well as dramatically improve performance. You can test it out with the `next` tag, and a short migration guide can be seen here: https://www.npmjs.com/package/next-mdx-remote/v/3.0.0-alpha.0#migrating-to-v3

```
npm install next-mdx-remote@next
```

Let me know if you've got feedback! You can feel free to DM me on Twitter ([@brkalow](https://twitter.com/brkalow)), or feel free to drop any feedback in the PR: https://github.com/hashicorp/next-mdx-remote/pull/121

Thanks 🙏

**Comment by BRKalow** [comment]
Thanks to everyone for the contributions and discussion here, we've just [released 3.0.0](https://github.com/hashicorp/next-mdx-remote/releases/tag/3.0.0) which should address this issue. 🙏
